apps/prairielearn/src/ee/lib/billing/plan-grants.ts

```diff
@@ -0,0 +1,58 @@
+import { type Response } from 'express';
+
+import { CourseInstanceSchema, InstitutionSchema, UserSchema } from '../../../lib/db-types';
+import { getEnrollmentForUserInCourseInstance } from '../../../models/enrollment';
+import {
+  getPlanGrantsForContextRecursive,
+  getPlanNamesFromPlanGrants,
+  getRequiredPlansForCourseInstance,
+} from '../../lib/billing/plans';
+import { planGrantsMatchPlanFeatures } from './plans-types';
+
+export async function checkPlanGrants(res: Response) {
+  // We'll only check plan grants for course instances, as students can't
+  // currently ever access a course directly. And even if they could, plan
+  // grans aren't associated with courses.
+  if (!res.locals.course_instance) {
+    return true;
+  }
+
+  // We won't check plan grants if the user has a specific role in the course
+  // or course instance. We always grant instructor-like users access to all
+  // features.
+  if (
+    res.locals.authz_data.course_role !== 'None' ||
+    res.locals.authz_data.course_instance_role !== 'None'
```

Comment:
Fwiw we have a sproc called `users_is_instructor_in_course` that does this check. Reading that sproc quickly makes it look like the check above is correct. Are we also checking for (global or institution) admins to give them automatic access?

Reply:
Maybe we can write a helper function that does the equivalent of the `users_is_instructor_in_course` sproc for use in TS code? Including the admin checks.

Reply:
This middleware will *always* run after `authzCourseOrInstance`, which will have already taken all of this into account (admins, "is this user an instructor", etc.). I see no reason to go back to the DB to ask for that information again.

Reply:
Good point, I was thinking about this in isolation.

Reply:
Comment about this added in c1994b5.